### PR TITLE
fix navigator bugs and clarify naming

### LINKED ISSUE #929: Fix navigator crash
### Description of the task
Navigator crashes due to exception and segfault, starting after #887. Test: visualizer with grsim backend. Anytime it navigates, it segfaults. Fix it. 

### Acceptance criteria

- [ ] Fix so it doesn't segfault
- [ ] Check that it doesn't segfault using visualizer
- [ ] Item C

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
Fixes bugs causing navigator to crash

### Testing Done
Visualizer

### Resolved Issues
Resolves #929

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Comment by jonathanlew** [comment]
Note that additional refactoring will be in future PRs

File: src/software/ai/navigator/navigator.cpp [hunk 1]
```diff
@@ -113,11 +113,13 @@ void Navigator::movePointNavigation(const MoveIntent &move_intent,
             LOG(WARNING) << "Path planner could not find a path";
             auto stop = std::make_unique<StopPrimitive>(move_intent.getRobotId(), false);
             current_primitive = std::move(stop);
+            break;
```
Comment (MathewMacDougall) [inline src/software/ai/navigator/navigator.cpp:116]: oh _LMAO_

File: src/software/gui/drawing/navigator.cpp [hunk 2]
```diff
@@ -22,11 +22,14 @@ AIDrawFunction drawNavigator(std::shared_ptr<Navigator> navigator)
 
         for (const auto& path : planned_paths)
         {
-            for (unsigned i = 0; i < path.size() - 1; i++)
+            if (path.size())
```
Comment (MathewMacDougall) [inline src/software/gui/drawing/navigator.cpp]: Why is the extra `if` necessary? Isn't this the same behavior as just the loop?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Reviewing first since I don't want people to have their navigator crash while testing, and because of everyone's presumably limited availability this long weekend.

Just a small nit. Curious why we didn't see these crashes at RoboCup?

**Comment by jonathanlew** [comment]
> 
> 
> Reviewing first since I don't want people to have their navigator crash while testing, and because of everyone's presumably limited availability this long weekend.
> 
> Just a small nit. Curious why we didn't see these crashes at RoboCup?

These are new from #887 I just didn't clean up a few things when everything moved around

File: src/software/ai/navigator/navigator.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/navigator/navigator.cpp:116]: yeah whoops...

File: src/software/gui/drawing/navigator.cpp (hunk 2 above)
Comment (jonathanlew) [reply src/software/gui/drawing/navigator.cpp]: The new API now has us returning empty vector for no path. This handles that here

Comment (MathewMacDougall) [reply src/software/gui/drawing/navigator.cpp]: Yea but if the vector is empty the loop wouldn't run anyway?

Comment (jonathanlew) [reply src/software/gui/drawing/navigator.cpp]: We want to check that path.size() is greater than 0

Comment (jonathanlew) [reply src/software/gui/drawing/navigator.cpp]: It turns out that without this change, it segfaults

Comment (garethellis0) [reply src/software/gui/drawing/navigator.cpp]: Just so we're being super explicit, could we make this:
``` suggestion
            if (path.size() > 0)
```

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Why didn't we catch this before? Seems like blatantly incorrect behavior, could we add a test that fails with the incorrect code and is fixed with this PR?

**Comment by jonathanlew** [comment]
> 
> 
> Why didn't we catch this before? Seems like blatantly incorrect behavior, could we add a test that fails with the incorrect code and is fixed with this PR?

@MathewMacDougall are there tests for the visualizer somewhere? I can't seem to find a navigator_test.cpp in drawing

**Comment by jonathanlew** [comment]
> 
> 
> Why didn't we catch this before? Seems like blatantly incorrect behavior, could we add a test that fails with the incorrect code and is fixed with this PR?

@MathewMacDougall and I were discussing this, and this unsigned comparison issue might have come out of enforcing compiler warnings

**Comment by MathewMacDougall** [comment]
@jonathanlew there's not a test for the drawing functions yet

**Comment by jonathanlew** [comment]
> 
> 
> @jonathanlew there's not a test for the drawing functions yet

@MathewMacDougall is there an issue filed for that?

**Comment by MathewMacDougall** [comment]
Part of #837 which is a superticket of Visualizer improvements right now.

**Review by MathewMacDougall — DISMISSED** [review]
LGTM

**Comment by garethellis0** [comment]
> > @jonathanlew there's not a test for the drawing functions yet
> 
> @MathewMacDougall is there an issue filed for that?

Apologies, realized I wasn't super clear. I mean why did the tests not fail for the navigator given that we were throwing exceptions in cases we shouldn't have been?

**Comment by jonathanlew** [comment]
Made new tests for navigator corner cases. Can I get a re-review @garethellis0 @MathewMacDougall

**Review by henryymliu — COMMENTED** [review]
Overall LGTM

File: src/software/ai/navigator/navigator_test.cpp [hunk 3]
```diff
@@ -201,3 +203,109 @@ TEST(NavigatorTest, convert_multiple_intents_to_primitives)
...
+    // tests
+
+    Timestamp current_time(Timestamp::fromSeconds(123));
+    Field field(0, 0, 0, 0, 0, 0, 0, current_time);
+    Ball ball(Point(1, 2), Vector(-0.3, 0), current_time);
+    Team friendly_team(Duration::fromMilliseconds(1000));
+    Team enemy_team(Duration::fromMilliseconds(1000));
+
+    // An arbitrary fixed point in time
+    // We use this fixed point in time to make the tests deterministic.
```
[the remark this answers is not in the document]
Comment (jonathanlew) [reply src/software/ai/navigator/navigator_test.cpp:273]: Yeah it's a bit of a mess but this will be cleaned up into the setup/constructor soon(tm)

File: src/software/ai/navigator/navigator_test.cpp [hunk 4]
```diff
@@ -201,3 +203,109 @@ TEST(NavigatorTest, convert_multiple_intents_to_primitives)
     auto pivot_primitive = dynamic_cast<PivotPrimitive &>(*(primitive_ptrs.at(1)));
     EXPECT_EQ(expected_pivot_primitive, pivot_primitive);
 }
+
+TEST(NavigatorTest, move_intent_with_one_point_path_test_path_planner)
+{
+    // TODO: refactor this into the setup and constructor and add more of these types of
+    // tests
```
Comment (garethellis0) [inline src/software/ai/navigator/navigator_test.cpp:210]: TODO?

Comment (jonathanlew) [reply src/software/ai/navigator/navigator_test.cpp:210]: Yeah it works but it needs a refactor when a complete set of tests are written for navigator.

**Review by garethellis0 — APPROVED** [review]
LGTM

**Review by MathewMacDougall — APPROVED** [review]
